### ISSUE TITLE
Bug report: infinite strings permitted

### ISSUE THREAD

**Issue by SapphireBrand** [polterguy/lizzie#42]
I think that the code must require strings to have a bounded length. Otherwise invalid or malicious input can crash the interpreter.

The trend in language design is for such things to be implementation defined. So I guess that it is possible for your C# implementation to say that it is permissible for strings to be large enough to crash your computer. :-)

I am making the TypeScript implementation enforce a limit that is settable by the environment and defaults to 65536.

**Comment by polterguy** [comment]
Hmm, interesting problem, although I am not sure what to say about it to be honest with you. In some ways, I kind of myself believe it should be up to the C# class to decide how long a string can be, which is debated [here](https://stackoverflow.com/questions/140468/what-is-the-maximum-possible-length-of-a-net-string).

From the other side of the argument, I realize that it would probably be _"unwise"_ to allow a dynamically interpreted language (Lizzie) to be allowed to crash its host process due to an `OutOfMemoryException` ...

I think I'll probably implement some sort of setting on the `Binder`, like I did with function invocations, allowing the consumer to decide the maximum length of strings, and explicitly set it if he or she wants to, and just allowing a `OutOfMemoryException` to occur if it's not explicitly changed by the host process on the `Binder` instance. The same way I did with `MaxStackSize` ...

First of all because I don't want to create (unnecessary) backwards compatibility problems, for people already using it to parser _"humongously large string based datasets"_, secondly because I want to allow for parsing very large strings, and not force consumers to not be able to do this if they really want to ...

Thoughts ...?

**Comment by polterguy** [comment]
Although in a way it doesn't really matter that much, since the _"real problem"_ in fact arguably is the size of the code (providing infinitely large Lizzie code, making the compiler choke), I still chose to implement an optional setting on the `LizzieTokeniser` that allows consumers to themselves declare what size they want to tolerate.

I did this since supplying the maximum size of Lizzie snippets is difficult (to say the least), since among other things the tokenizing logic allows for supplying an enumerable of streams, which might not be possible to determine the size of before starting the tokenizing process, since there are no requirements to that the streams are searchable, and might be _"forward only"_ streams - Which is an important feature to be allowed to tokenize code that is for instance supplied directly to the Lizzie compiler over e.g. a network socket, etc ...

**Comment by SapphireBrand** [comment]
Good point about infinitely large Lizzie code.

I put a limit on string size into the TypeScript Lizzie, but you are right that a limit on code size will cover a much larger range of situations.

A Lizzie program of small size can be created that will consume arbitrarily large amounts of process memory, but that is a separate issue, which I will file a report for so you can think that over separately.
